Make it possible to set WeasyPrint options

dekkers opened the pull request:
WeasyPrints can be given an options dict in the `render` and `write_pdf` methods. This PR makes it possible to do that with django-weasyprin by setting the `pdf_options` attribute on the view class or by implementing the `get_pdf_options` method.

liZe commented on django_weasyprint/views.py:
```diff
@@ -82,6 +83,7 @@ def get_document(self):
         return html.render(
             stylesheets=self.get_css(base_url, url_fetcher, font_config),
             font_config=font_config,
+            option=self._options,
```
Shouldn’t we pass options as separate keyword arguments? The [API](https://doc.courtbouillon.org/weasyprint/stable/api_reference.html#weasyprint.HTML.render) asks for `**options`, not `option`. Something like `**self._options` instead of `option=self._options` is probably appropriate.

fdemmer replied:
thank you for your comments! i think you accidentally pasted the wrong link, @liZe ;)

API: https://doc.courtbouillon.org/weasyprint/stable/api_reference.html#weasyprint.HTML.render

yes, if we want to pass the "options" all at once, the kwarg should definitely be called `options`. it _is_ a dict of potentially _multiple_ options.

however, that may be cause for confusion/bugs, as we already provide a `stylesheets` kwarg, which actually _is_ a [DEFAULT_OPTIONS](https://doc.courtbouillon.org/weasyprint/stable/api_reference.html#weasyprint.DEFAULT_OPTIONS) parameter.

a good way around this may be using [`setdefault`](https://docs.python.org/3/library/stdtypes.html#dict.setdefault) to add `stylesheets` to `self._options` before passing it to `render()` using **-unpacking. maybe a copy of the options dict is a good idea (similar to [weasyprint's `render()`](https://github.com/Kozea/WeasyPrint/blob/4edb03d9056093bd6cfbc1cb2613bd5c584f957b/weasyprint/__init__.py#L192)).

liZe replied:
> thank you for your comments! i think you accidentally pasted the wrong link, @liZe ;)

🤣

liZe commented on django_weasyprint/views.py:
```diff
@@ -90,7 +92,7 @@ def rendered_content(self):
         """
         document = self.get_document()
-        return document.write_pdf()
+        return document.write_pdf(options=self._options)
```
Same as above.

fdemmer replied:
yes, should use **; no need for copy/setdefault, i think, since we do not provide any other kwargs here (valid kwargs other than `options` here are: [`target`, `zoom`, `finisher`](https://doc.courtbouillon.org/weasyprint/stable/api_reference.html#weasyprint.document.Document.write_pdf)).

liZe reviewed (changes requested):
I like the overall idea.

I can’t comment the Django part (@fdemmer can provide better feedback on this point!), but there are some details to change regarding WeasyPrint’s API in order to make this PR work correctly.

fdemmer reviewed (changes requested):
Thank you for the suggestion and contribution, @dekkers ! :heart:

Please add a test, where the new `pdf_options` attribute of `WeasyTemplateResponseMixin` is used and make sure it still works, with `pdf_stylesheets` provided aswell.

An update to the example in the README would also be much apprechiated.

dekkers committed:
Make it possible to set WeasyPrint options

dekkers commented:
Thanks for the feedback. Took me a while to find the time to properly fix this, but I've fixed the passing of the options and used  `setdefault` to set stylesheets as suggested. I have also added a unit test that checks whether the stylesheets still work with `pdf_options` given and checks whether the generated PDF has the PDF version specified in the `pdf_options`.

hveis commented:
This is brilliant!! An awesome fix
I was looking into the API and wondered how this wasn't implemented/documented for django-weasyprint 
Hoping this pull request gets merged soon 
Thanks to all that helped
